homeassistant/components/livisi/switch.py

```diff
@@ -0,0 +1,166 @@
+    room_name: str = coordinator.rooms[room_id]
+    livisi_switch = LivisiSwitch(
+        config_entry,
+        coordinator,
+        unique_id=device["id"],
+        manufacturer=device["manufacturer"],
+        device_type=device["type"],
+        name=config_details["name"],
+        capability_id=capabilities[0],
+        room=room_name,
+    )
+    return livisi_switch
+
+
+class LivisiSwitch(CoordinatorEntity[LivisiDataUpdateCoordinator], SwitchEntity):
+    """Represents the Livisi Switch."""
+
+    def __init__(
+        self,
+        config_entry: ConfigEntry,
+        coordinator: LivisiDataUpdateCoordinator,
+        unique_id: str,
+        manufacturer: str,
+        device_type: str,
+        name: str,
+        capability_id: str,
+        room: str,
+    ) -> None:
+        """Initialize the Livisi Switch."""
+        self.config_entry = config_entry
+        self._attr_unique_id = unique_id
+        self._attr_name = name
+        self._capability_id = capability_id
+        self.aio_livisi = coordinator.aiolivisi
+        self._attr_available = False
+        self._attr_device_info = DeviceInfo(
+            identifiers={(DOMAIN, unique_id)},
+            manufacturer=manufacturer,
+            model=device_type,
+            name=name,
+            suggested_area=room,
+            via_device=(DOMAIN, config_entry.entry_id),
+        )
+        super().__init__(coordinator)
+
+    async def async_turn_on(self, **kwargs: Any) -> None:
+        """Turn the entity on."""
+        response = await self.aio_livisi.async_pss_set_state(
+            self._capability_id, is_on=True
+        )
+        if response is None:
+            self._attr_available = False
+            raise HomeAssistantError(f"Failed to turn on {self._attr_name}")
+
+    async def async_turn_off(self, **kwargs: Any) -> None:
+        """Turn the entity off."""
+        response = await self.aio_livisi.async_pss_set_state(
+            self._capability_id, is_on=False
+        )
+        if response is None:
+            self._attr_available = False
+            raise HomeAssistantError(f"Failed to turn off {self._attr_name}")
+
+    async def async_added_to_hass(self) -> None:
+        """Register callbacks."""
+        response = await self.coordinator.async_get_pss_state(self._capability_id)
+        if response is None:
+            self._attr_is_on = False
+            self._attr_available = False
+        else:
+            self._attr_is_on = response
+        self.async_on_remove(
+            async_dispatcher_connect(self.hass, LIVISI_STATE_CHANGE, self.update_states)
+        )
+        self.async_on_remove(
+            async_dispatcher_connect(
+                self.hass, LIVISI_REACHABILITY_CHANGE, self.update_reachability
+            )
+        )
+
+    @callback
+    def update_states(self, device_id_state) -> None:
+        """Update the states of the switch device."""
+        if device_id_state[ID] != self._capability_id:
```

Comment:
Consider including the ID in the signal string the the dispatcher uses. Then we can remove the checks here in the callbacks and the dispatch will be more efficient.

Reply:
I cannot use the ID of the device here. In the LIVISI app, the update of the device states is sent using the capability ID, not the device ID which is stored in HomeAssistant (In LIVISI API we have device ID and capability ID, the capability is what a device can do, and the state changes are being sent on the WebSocket using the capability ID).

Reply:
I can use the device ID when I update the device's reachability because the reachability state is being sent with the device ID, which is stored in HomeAssistant as a unique ID.

Reply:
We know `capability_id` both when we send the dispatch signal with `async_dispatcher_send` and when we connect the dispatcher with `async_dispatcher_connect`. My suggestion is to include the `capability_id` in the signal string, eg `f"{LIVISI_STATE_CHANGE}_{self._capability_id}"`.

We can do the same for the update_reachability signal using the device id.

Reply:
Thank you for your fast answer. I updated the code.